sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,855 @@
+	}
+	else {
+		*(sdio_common.base + reg_sys_ctrl) = SDHC_SYSCTL_CLK_DIV_8 | SDHC_SYSCTL_RESERVED;
+	}
+
+	/* Send 80 initialization clock cycles
+	 * to the card, and wait until it finishes.
+	 * Should be done in ~3us on 25MHz clock */
+	*(sdio_common.base + reg_sys_ctrl) |= (1 << 27);
+	for (i = 0; i < SDHC_RETRIES; i++) {
+		val = *(sdio_common.base + reg_sys_ctrl) & (1 << 27);
+		if (val == 0) {
+			break;
+		}
+		usleep(10);
+	}
+	if (val != 0) {
+		return -1;
+	}
+
+	return 0;
+}
+
+
+void sdio_free(void)
+{
+	sdhc_reset(SDHC_SYSCTL_RESET_ALL);
+
+	/*
+	 * FIXME: the following resources:
+	 * isrHandle, eventLock, eventCond
+	 * are not destroyed since they are used
+	 * by running event service thread.
+	 * The same applies for hardware register
+	 * base address which remains mapped.
+	 * this is due to supposed problems with
+	 * threadJoin function which would allow
+	 * to stop the thread properly.
+	 */
+
+	sdio_common.sdioInitialized = 0;
+	sdio_common.blocksz = 0;
+
+	if (sdio_common.cmdLock >= 0) {
+		resourceDestroy(sdio_common.cmdLock);
+		sdio_common.cmdLock = (handle_t)NULL;
+	}
+	if (sdio_common.dmaptr != NULL) {
+		munmap(sdio_common.dmaptr, DMA_BUFFER_SIZE);
+		sdio_common.dmaptr = NULL;
+		sdio_common.dmaphys = (addr_t)NULL;
+	}
+
+	*(sdio_common.base + reg_int_status_en) = 0;
+	*(sdio_common.base + reg_int_signal_en) = 0;
+
+	memset(&sdio_common.eventHandlers, 0, sizeof(sdio_common.eventHandlers));
+	memset(&sdio_common.eventHandlerArgs, 0, sizeof(sdio_common.eventHandlerArgs));
+}
+
+
+int sdio_transferDirect(sdio_dir_t dir, uint32_t address, uint8_t area, uint8_t *data)
+{
+	int rslt = 0;
+	uint32_t response, arg = 0;
+
+	arg |= ((address & 0x1ffff) << 9);
+	arg |= ((area & 0x7) << 28);
+	arg |= (dir << 31);
+
+	/* On direct write transfers data is part of argument */
+	if (dir == sdio_write) {
+		arg |= *data;
+	}
+
+	rslt = sdio_cmdSend(SDIO_CMD_RW_DIRECT, arg, &response);
+
+	/* Data is part of response on read transfers
+	 * and also on writes with RaW (read after write)
+	 * flag set - SDIO simplified spec v3.00, page 36 */
+	if (dir == sdio_read) {
+		*data = response & 0xff;
+	}
+
+	return rslt;
+}
+
+
+static int _sdio_transferBulk(sdio_dir_t dir, int blockMode, uint32_t address,
+	uint8_t area, uint8_t *data, size_t len)
+{
+	uint32_t val, cmd, mix, blk;
+	uint32_t arg = 0;
+	uint16_t count;
+
+	if (blockMode == 1 && (sdio_common.blocksz == 0 || len % sdio_common.blocksz != 0)) {
+		printf("libsdio: transfer size is not a multiple of block size\n");
+		return EINVAL;
+	}
+	count = (blockMode) ? (len / sdio_common.blocksz) : len;
+
+	/* construct argument */
+	arg |= (dir << 31);
+	arg |= ((area & 0x7) << 28);
+	arg |= ((address & 0x1ffff) << 9);
+	arg |= (count & 0x1ff);
+	arg |= ((blockMode & 0x1) << 27);
+	arg |= (1 << 26);
+
+	val = *(sdio_common.base + reg_pres_state);
+	if (val & 0x7) {
+		printf("libsdio: data interface is busy (pres_state=%x)\n", val);
+		return -1;
```

Comment:
Maybe to make this code easier for debug use errno error codes in all `public` API functions. So here you could return `-EBUSY`. Check this list of error codes: 
https://pubs.opengroup.org/onlinepubs/7908799/xsh/errors.html

Reply:
This part now returns `-EBUSY`.